contracts/src/core/libraries/AccessControlLib.sol

```diff
@@ -0,0 +1,83 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.12;
+
+import {EnumerableSet} from "lib/openzeppelin-contracts/contracts/utils/structs/EnumerableSet.sol";
+
+library AccessControlStorage {
+    struct Layout {
+        mapping(bytes32 => AccessControlLib.RoleData) roles;
+    }
+
+    string internal constant STORAGE_ID = "access.control.storage";
+    bytes32 internal constant STORAGE_POSITION =
+        keccak256(abi.encode(uint256(keccak256(abi.encodePacked(STORAGE_ID))) - 1)) & ~bytes32(uint256(0xff));
+
+    function layout() internal pure returns (Layout storage s) {
+        bytes32 position = STORAGE_POSITION;
+        assembly {
+            s.slot := position
+        }
+    }
+}
+
+library AccessControlLib {
```

Comment:
Can we document this? In plain english how does it work? What capabilities/features does it offer? How does it differ from basic OZ accessControl that most people would expect. etc